Tests/OpenCombineTests/PublisherTests/MapTests.swift

```diff
@@ -0,0 +1,91 @@
+//
+//  MapTests.swift
+//
+//
+//  Created by Anton Nazarov on 25/06/2019.
+//
+
+import XCTest
+
+#if OPENCOMBINE_COMPATIBILITY_TEST
+import Combine
+#else
+import OpenCombine
+#endif
+
+@available(macOS 10.15, *)
+final class MapTests: XCTestCase {
+
+    static let allTests = [
+        ("testJust", testJust),
+        ("testEmpty", testEmpty),
+        ("testError", testError),
+        ("testRange", testRange)
+    ]
+
+    func testJust() {
+        // Given
+        let expectedEvent = 1
+        let tracking = TrackingSubscriberBase<Never>(receiveSubscription: { $0.request(.unlimited) })
+        let map = Publishers.Just(expectedEvent).map { $0 * 2 }
+        // When
+        map.subscribe(tracking)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .value(expectedEvent * 2),
+            .completion(.finished)
+        ])
+    }
+
+    func testEmpty() {
+        // Given
+        let tracking = TrackingSubscriberBase<Never>()
+        _ = Publishers.Empty().map { $0 * 2 }.subscribe(tracking)
+        // When
+        tracking.subscriptions.first?.request(.unlimited)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .completion(.finished)
+        ])
+    }
+
+    func testError() {
+        // Given
+        let expectedError = TestingError.oops
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.unlimited) })
+        let map = Publishers.Fail(error: expectedError).map { $0 * 2 }
+        // When
+        map.subscribe(tracking)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .completion(Subscribers.Completion<TestingError>.failure(expectedError))
+        ])
+    }
+
+    func testRange() {
```

Comment:
This is a good test!

If you have time, it'd be great to also add tests that ensure that demand is properly handled — something like this for `DropWhile`:
https://github.com/broadwaylamb/OpenCombine/blob/17f9b0a29975106d2f51d5d9b7858c5b1e612019/Tests/OpenCombineTests/PublisherTests/DropWhileTests.swift#L90-L129

Note that in the DropWhile case the request-demand logic was probably different from Map.

You may want to steal some more tests from the DropWhile suite :)

Reply:
Sure. Added few more. Please review